tests/basis_test.py

```diff
@@ -3,328 +3,296 @@
-    numpy.random.seed(567)
-    xs = numpy.random.rand(20) * 2 - 1
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    fn = ChebyshevFirstKind(5)
-    assert numpy.allclose(f0(xs),special.eval_chebyt(0,xs))
-    assert numpy.allclose(f1(xs),special.eval_chebyt(1,xs))
-    assert numpy.allclose(fn(xs),special.eval_chebyt(5,xs))
 
 
-def test_cheb_derivative():
-    """Test if the correct derivative is generated."""
-    f0 = ChebyshevFirstKind(0)
-    f1 = ChebyshevFirstKind(1)
-    f2 = ChebyshevFirstKind(2)
-    assert f0.derivative(1) == pytest.approx(0)
-    assert f1.derivative(1) == pytest.approx(1)
-    assert f2.derivative(1) == pytest.approx(4)
-    
-    assert f0.derivative(-0.5) == pytest.approx(0)
-    assert f1.derivative(-0.5) == pytest.approx(1)
-    assert f2.derivative(-0.5) == pytest.approx(-2)
-
-    assert numpy.isclose(f0.derivative([1,-0.5]),[0,0]).all()
-    assert numpy.isclose(f1.derivative([1,-0.5]),[1,1]).all()
-    assert numpy.isclose(f2.derivative([1,-0.5]),[4,-2]).all()
-
-
-def test_cheb_2nd_derivative():
-    """Test if the correct derivative is generated."""
-    u0 = ChebyshevSecondKind(0)
-    u1 = ChebyshevSecondKind(1)
-    u2 = ChebyshevSecondKind(2)
-    assert u0.derivative(1) == pytest.approx(0)
-    assert u1.derivative(1) == pytest.approx(2)
-    assert u2.derivative(1) == pytest.approx(8)
+def test_cheb_initial():
+    """test degrees returns correctly"""
+    f2 = smolyay.basis.ChebyshevFirstKind(2)
+    assert f2.n == 2
+    assert numpy.array_equal(f2.domain, [-1, 1])
 
-    assert u0.derivative(-1) == pytest.approx(0)
-    assert u1.derivative(-1) == pytest.approx(2)
-    assert u2.derivative(-1) == pytest.approx(-8)
 
-    assert u0.derivative(0.5) == pytest.approx(0)
-    assert u1.derivative(0.5) == pytest.approx(2)
-    assert u2.derivative(0.5) == pytest.approx(4)
+def test_cheb_ntype():
+    """test type of degree"""
+    f2 = smolyay.basis.ChebyshevFirstKind(2)
+    assert isinstance(f2.n, int)
 
-    assert numpy.isclose(u0.derivative([1,-1,0.5]),[0,0,0]).all()
-    assert numpy.isclose(u1.derivative([1,-1,0.5]),[2,2,2]).all()
-    assert numpy.isclose(u2.derivative([1,-1,0.5]),[8,-8,4]).all()
 
+def test_cheb_n_setter():
+    """test degree setter"""
+    f2 = smolyay.basis.ChebyshevFirstKind(2)
+    f2.n = 3
+    assert f2.n == 3
+    assert isinstance(f2.n, int)
+
+
+@pytest.mark.parametrize("n", list(range(20)))
+def test_cheb_call_random_points(n):
+    """Test chebyshev polynomial at some degree at some input"""
+    numpy.random.seed(567)
+    xs = numpy.random.rand(20) * 2 - 1
+    f = smolyay.basis.ChebyshevFirstKind(n)
+    assert numpy.allclose(f(xs), special.eval_chebyt(n, xs))
+
+
+def test_cheb_call_extrema_points():
+    """Test chebyshev polynomial at some degree at some input"""
+    extrema_points = [-1.0, -1 / numpy.sqrt(2), 0, 1 / numpy.sqrt(2), 1]
+    extrema_output = [1, -1, 1, -1, 1]
+    f = smolyay.basis.ChebyshevFirstKind(4)
```

Comment:
It is easier to read this with the function coming first, since then we know why these are the extrema.
```suggestion
    f = smolyay.basis.ChebyshevFirstKind(4)
    extrema_points = [-1.0, -1 / numpy.sqrt(2), 0, 1 / numpy.sqrt(2), 1]
    extrema_output = [1, -1, 1, -1, 1]
```